go-from-rust.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Early Impressions of Go from a Rust Programmer
+author: ['Nick Cameron']
+date: 2020-02-19
+summary:  Read this post to learn more. 
+tags: ['Engineering']
+categories: ['Engineering']
+image: 
+---
+
+# Early Impressions of Go from a Rust Programmer
+
+I've been using Go for the past few weeks. It's my first time using Go for a large (-ish), serious project. I've previously looked at Go a lot and played with examples and toy programs when researching features for Rust. Real programming is a very different experience.
+
+I thought it would be interesting to write about my impressions. I'll try and avoid too much comparison with Rust, but since that is the language I'm coming from, there will be some. I should declare in advance a strong bias for Rust, but I'll do my best to be objective.
+
+## General impression
+
+Programming with Go is nice. It had everything I wanted in the libraries and there were not too many rough edges. Learning it was a smooth experience, and it is a well-designed and practical language. Some examples: once you know the syntax, many idioms from other languages carry over to Go. Once you've learned some Go, it is relatively easy to predict other features. With some knowledge of other languages, I was able to read Go code and understand it without too much googling.
+
+It is a lot less frustrating and a lot more productive than using C/C++, Java, Python, etc. It did feel, however, like part of that generation of languages. It has learnt some lessons from them, and I think it is probably the best language of that generation; but it is definitely part of that generation. It is an incremental improvement, rather than doing something radically different (to be clear, this is not a value judgement, incremental is often good in software engineering). A good example of this is `nil`: languages like Rust and Swift have removed the concept of `null` and eliminated a whole class of errors. Go renames it to `nil` (why? No idea) and makes it a bit less dangerous (no null *values*, distinguishes between `nil` and `0`). But the core idea is still there, and so is the common runtime error of dereferencing a null pointer.
```

Comment:
Ruby uses nil and I believe it came from Lisp, so I wouldn't say they are renaming anything, just using a name that does already have a tradition and mean the same thing.